### PR TITLE
CRD CEL Validation: per-CRD estimated cost limit enforcement

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR is a part of https://github.com/kubernetes/kubernetes/issues/107573.

This enforces per-CRD estimated cost limits. It tracks the per-rule costs that contribute most significantly to the total cost and attributes the cost to them so developers may more easily track down which rules they need to optimize to resolve the issue.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

/sig api-machinery
/priority important-soon

### CODE REVIEW

**Comment by dims** [comment]
@liggitt this looks ready (CI jobs are green)

**Comment by liggitt** [comment]
this needs restacking on updated https://github.com/kubernetes/kubernetes/pull/108612 once that fixes the unit test issue

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 1]
```diff
@@ -742,16 +762,133 @@ func validateCustomResourceDefinitionValidation(ctx context.Context, customResou
...
+		c.totalCost = math.MaxUint64
+	} else {
+		c.totalCost += cost
+	}
+
+	if cost < StaticEstimatedCRDCostLimit/100 { // ignore rules that contribute < 1% of total cost
+		return
+	}
+	c.mostExpensive = append(c.mostExpensive, ruleCost{path: path, cost: cost})
+	sort.Slice(c.mostExpensive, func(i, j int) bool {
```
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:796]: Alternatively, I could do pure insertion sort on this (bubble in each new entry). For a size 4 list, I suspect it won't matter.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 2]
```diff
@@ -760,14 +774,45 @@ type costInfo struct {
 	// that the parent schemas offer no bound to the number of times a data element for the current
 	// schema can exist.
 	MaxCardinality *uint64
+	CRDCost        *crdCost
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:777]: document the meaning of this field... IIUC, this is a pointer in order to propagate and modify a shared value as we descend, right?

Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: I've added the guards to be super safe

Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: I've fleshed out the documentation as well, with notes about how the field lifecycle.

**Comment by jpbetz** [comment]
Squashed and rebased.

**Pull request merged**